Feature/detect format function

cleong110 opened the pull request:
DRAFT PR:

Adding a function to fix #142. Still TODO: 

- [x] go through the file and replace manual checks with this function. 
- [x] do some more testing to make sure there are no side effects

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -7,7 +7,36 @@
 
+SupportedPoseFormat = Literal["holistic", "openpose", "openpose_135"]
+
+def detect_pose_format(pose: Pose) -> SupportedPoseFormat:
```
I would suggest `detect_known_pose_format` (because technically you can use the library for arbitrary poses)

cleong110 commented:
OK, I've updated throughout `generic.py` to use the new `detect_known_format` function. I also added some pytests in, though not comprehensively.

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -28,41 +81,59 @@ def pose_hide_legs(pose: Pose):
+            f"Unsupported pose header schema {known_pose_format} for {pose_hide_legs.__name__}: {pose.header}"
+        )
     else:
-        raise ValueError("Unknown pose header schema for hiding legs")
```
will never get here, `detect_known_pose_format` would through before

cleong110 replied:
That's what it should do, I guess I kept it in superstitiously just in case!

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -1,12 +1,64 @@
+    raise ValueError(f"Could not get component_names from {pose_or_header_or_components}")
+
+
+def detect_known_pose_format(component_names: List[str]) -> SupportedPoseFormat:
```
i think maybe the input should be a `Pose` object or `PoseHeader`. usage is clunky, always:
```py
detect_known_pose_format(get_component_names(pose))
```

cleong110 replied:
Good idea! And I can use `get_component_names` inside the function.

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -1,12 +1,64 @@
+# from pose_format.utils.holistic import holistic_components
+# The import above creates an error: ImportError: Please install mediapipe with: pip install mediapipe
+
+SupportedPoseFormat = Literal["holistic", "openpose", "openpose_135"]
```
Given the function name, I'd change that to `KnownPoseFormats`. especially since OpenPose 135 is known but not really supported

cleong110 replied:
Good idea! Much better name

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -28,41 +81,59 @@ def pose_hide_legs(pose: Pose):
+        raise NotImplementedError(
+            f"Unsupported pose header schema {known_pose_format} for {pose_shoulders.__name__}: {pose_header}"
+        )
+    raise ValueError(f"Unknown pose header schema for {pose_shoulders.__name__}: {pose_header}")
```
will never get here. `detect_known` will through error

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -28,41 +81,59 @@ def pose_hide_legs(pose: Pose):
+        raise NotImplementedError(
+            f"Unsupported pose header schema {known_pose_format} for {hands_indexes.__name__}: {pose_header}"
+        )
+    raise ValueError(f"Unknown pose header schema for {hands_indexes.__name__}: {pose_header}")
```
will never get here

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -71,27 +142,32 @@ def pose_normalization_info(pose_header: PoseHeader):
+        raise NotImplementedError(
+            f"Unsupported pose header schema '{known_pose_format}' for {hands_components.__name__}: {pose_header}"
+        )
+    raise ValueError(f"Unknown pose header schema for {hands_components.__name__}: {pose_header}")
```
will never get here

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -71,27 +142,32 @@ def pose_normalization_info(pose_header: PoseHeader):
-                                               p2=(component_name, line[1]))
-    normalizer = PoseNormalizer(plane=plane, line=line)
+    plane_info = hand_pose.header.normalization_info(
+        p1=(component_name, plane[0]), p2=(component_name, plane[1]), p3=(component_name, plane[2])
```
new line for each, would be nicer

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -107,7 +183,29 @@ def normalize_hands_3d(pose: Pose, left_hand=True, right_hand=True):
+            return holistic_utils.holistic_components()
+        except ImportError:
+            # TODO: test this branch
+            test_file_path = Path("tests/data/mediapipe.pose")
```
i'm against. if someone wants holistic components, they should have mediapipe installed.

cleong110 replied:
I think that makes sense, and we can set it up that way, I guess my question then is: what about the goal of keeping the library lightweight as discussed in #111? 

Do we try to either (1) load/parse/manipulate various known formats without creating dependencies (as in [`reduce_holistic`](https://github.com/sign-language-processing/pose/blob/master/src/python/pose_format/utils/generic.py#L166), or (2) accept the mediapipe dependency in certain cases?

Currently the library seems to do some of both - the OpenPose formats provide their components as a List, whereas holistic requires installing the library and using a function.

cleong110 replied:
I will simply re-raise the importerror for now I suppose.

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -122,21 +220,29 @@ def fake_pose(num_frames: int, fps=25, dims=2, components=OpenPose_Components):
+        raise NotImplementedError(
+            f"{known_pose_format} pose header schema unsupported for get_hand_wrist_index: {pose.header}"
+        )
+    raise ValueError(f"Unknown schema for {get_hand_wrist_index}: {pose.header}")
```
never?

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -122,21 +220,29 @@ def fake_pose(num_frames: int, fps=25, dims=2, components=OpenPose_Components):
+        raise NotImplementedError(
+            f"{known_pose_format} pose header schema unsupported for {get_body_hand_wrist_index.__name__}"
+        )
+    raise ValueError(f"Unknown pose header schema for {get_body_hand_wrist_index.__name__} {pose.header}")
```
never?

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -175,27 +282,152 @@ def reduce_holistic(pose: Pose) -> Pose:
     """
     # To avoid installing mediapipe, we just hardcode the face contours given the above code
     face_contours = [
-        '0', '7', '10', '13', '14', '17', '21', '33', '37', '39', '40', '46', '52', '53', '54', '55', '58', '61', '63',
```
i don't like this change

cleong110 replied:
Ah, `black` changed that.

cleong110 replied:
I think I need to just stop using that, we keep running into issues like this even if I configure it to have a longer line length.

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -175,27 +282,152 @@ def reduce_holistic(pose: Pose) -> Pose:
+        "ANKLE",
+        "HEEL",
+        "FOOT_INDEX",
+    ]  # Face  # Hands  # Feet
```
bad change - they were correctly clustered, and now the comments don't make sense.

cleong110 replied:
Reverted

cleong110 commented:
OK, ready for the next round, I believe. 
* detect function takes a pose or poseheader now
* various formatting changes.
* added some unknown-format poses to the tests, detect_known_format correctly raises an exception.

AmitMY reviewed (approved):
Very cool, thank you :)

cleong110 commented:
Taking a crack at fixing the tests. They run on my machine with python 3.10, but crash when run on github on 3.8. Why?

cleong110 commented:
I'm kinda stumped by this pose import circle. Which I think I introduced actually, with the "quick and simple" PR where pose.normalize() would call it by default.

Pose imports utils.generic to get pose_normalization_info
utils imports Pose
^ thus a circle. 

Where I'm stuck:
* Simplest solution might be to remove pose_normalization_info import from Pose
* but we do want to have that feature, so do we move the implementation into Pose? 
* But then pose_shoulders() needs to move too...

AmitMY commented:
Fixed the circular issue at https://github.com/sign-language-processing/pose/commit/56e67179c9bc2da8bf2da589ae5a2404a2bea2df

cleong110 commented:
> Fixed the circular issue at [56e6717](https://github.com/sign-language-processing/pose/commit/56e67179c9bc2da8bf2da589ae5a2404a2bea2df)

That'll do it!

cleong110 commented:
OK, latest problem is with the pipe type annotations, which prior to 3.10 will throw errors. 

https://peps.python.org/pep-0604/

Tests for versions 3.10-3.12 are currently expected to fail with NotImplementedErrors. We can either leave them failing as a reminder to implement, or just add a with pytest.raises to catch those.

AmitMY commented:
I susggest we use Union instead of the pipe, to support older versions

cleong110 commented:
While trying to get tests to pass, I discovered two problems with correct_wrist:
1. It was modifying the input Pose's data despite returning a Pose object. Added a deepcopy operation to prevent this. 
3. It was creating stacked_conf with shape 3 points (XYZ), even if the format called for 2 (XY). Changed out the hardcoded "3" to check the shape of the data, thus:

```
point_coordinate_count = wrist.shape[-1]
    stacked_conf = np.stack([wrist_conf] * point_coordinate_count, axis=-1) 
```

cleong110 commented:
All tests now green on my machine!

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -17,7 +17,7 @@
 
 def get_component_names(
-    pose_or_header_or_components: Pose | PoseHeader | List[PoseHeaderComponent] | List[str],
+    pose_or_header_or_components: Union[ Pose , PoseHeader , List[PoseHeaderComponent] , List[str]],
```
too much spacing
`Union[Pose,PoseHeader,List[PoseHeaderComponent],List[str]],`

but also, why is anything necessary except for `Pose` or `PoseHeader`? too many options imo

cleong110 replied:
This function I envisioned as the "whatever type it is, don't care what, give me the component names". Which seemed useful to me, I find that this library often involves things like 

```python
component_names = [c.name for c in some_list_of_components]
```

Or 

```python
component_names = [c.name for c in some_pose_object.header.components]
```

Or 

```python
component_names = [c.name for c in some_pose_header.components]
```

Which are all basically the same thing. 

I almost feel like get_component_names() should be a member of Pose or PoseHeader actually, it's so common

cleong110 replied:
But it's a bit out of scope for this PR, so let's just go with Pose and PoseHeader as the only options

AmitMY commented on src/python/pose_format/utils/generic.py:
```diff
@@ -245,13 +253,14 @@ def correct_wrist(pose: Pose, hand: str) -> Pose:
     pose.body.data[:, :, body_wrist_index] = new_wrist_data
     pose.body.confidence[:, :, body_wrist_index] = new_wrist_conf
-    return pose
+    return pose  
```
why this spacing?

cleong110 replied:
No clue!

AmitMY reviewed (approved):
great! thanks for the hard work :)
